### PR TITLE
Fixed Full Host image generation on 1.13

### PR DESCRIPTION
**Pull request by lzap**
PXELoader flag in 1.13 broke Full Host generation as the parameter is not
passed. We can fix this here, or I can provide default parameter in core to
remain compatibility (but it can bite). I prefer being explicit.

Anyway, this is bad regression, any chance to squeeze this into 1.13 plugin? Do
you want me to do a ticket?

### CODE REVIEW

**Comment by domcleal** [comment]
Can you add a unit test of `generate_full_host` please? It's OK to stub/expect `generate` itself.

**Comment by lzap** [comment]
Sure, I had to slightly refactor the helper in order to use it outside controller context.

File: test/unit/iso_generator_test.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  setup :setup_org_loc
+  setup :setup_subnet
+  setup :setup_host
+
+  setup do
+    @host.build = true
+  end
+
+  test "full host image generation generates via PXELinux type" do
+    @host.expects(:generate_pxe_template).with(:PXELinux)
```
Comment (domcleal) [inline test/unit/iso_generator_test.rb]: Doesn't stubbing this undermine the purpose of the test, to make sure it actually works? If it's stubbed expecting an argument and the method signature changes, the test will never fail.

Comment (lzap) [inline test/unit/iso_generator_test.rb]: Well my goal was to test only PXELinux option in `generate_pxe_template`. Anyway, if I understand you correctly, you want this test to be complete so changed it accordingly.

Also added two more tests on the filename part, which looks quite fragile.

File: test/unit/iso_generator_test.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    @host.operatingsystem.expects(:pxe_files).returns(boot_files)
+    ForemanBootdisk::ISOGenerator.expects(:generate).with({
+      :isolinux => 'Fake kernel line BOOT/NUXATOMIC_7_3_X86_64_VMLINUZ',
+      :files => [[['BOOT/NUXATOMIC_7_3_X86_64_VMLINUZ', '/tmp/pxeboot/vmlinuz']],
+        [['BOOT/ATOMIC_7_3_X86_64_INITRD_IMG', '/tmp/pxeboot/initrd.img']]]}, anything)
+    ForemanBootdisk::ISOGenerator.generate_full_host(@host)
+  end
+
+  test "full host image generation generates via PXELinux type" do
+    @host.expects(:generate_pxe_template).with(:PXELinux).returns("Fake content")
```
Comment (domcleal) [inline test/unit/iso_generator_test.rb]: > What exactly will cause the test to fail? Are you commenting on line 15 or 17?

This was on line 15, before you changed it. I'm saying that a change in the `generate_pxe_template` signature will _not_ cause the test to fail. This shouldn't be stubbed, it should be calling the real method.

File: test/unit/iso_generator_test.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  setup :setup_bootdisk
+  setup :setup_org_loc
+  setup :setup_subnet
+  setup :setup_host
+
+  setup do
+    @host.build = true
+  end
+
+  test "generate_full_host creates with ISO-compatible file names" do
```
Comment (domcleal) [inline test/unit/iso_generator_test.rb]: These look unrelated to the fix, please move them into their own PR.

**Comment by lzap** [comment]
Rebased. Removed the two guys:

```ruby
  test "generate_full_host creates with ISO-compatible file names" do
    @host.expects(:generate_pxe_template).with(:PXELinux).returns("Fake kernel line boot/Debian-8.1-x86_64-vmlinuz")
    boot_files = [
      {:"boot/Debian-8.1-x86_64"=>"/tmp/pxeboot/vmlinuz"},
      {:"boot/Debian-8.1-x86_64"=>"/tmp/pxeboot/initrd.img"}
    ]
    @host.operatingsystem.expects(:pxe_files).returns(boot_files)
    ForemanBootdisk::ISOGenerator.expects(:generate).with({
      :isolinux => 'Fake kernel line BOOT/DEBIAN_8_1_X86_64_VMLINUZ',
      :files => [[['BOOT/DEBIAN_8_1_X86_64_VMLINUZ', '/tmp/pxeboot/vmlinuz']],
      [['BOOT/DEBIAN_8_1_X86_64_INITRD_IMG', '/tmp/pxeboot/initrd.img']]] }, anything)
    ForemanBootdisk::ISOGenerator.generate_full_host(@host)
  end

  test "generate_full_host creates with ISO-compatible long file names" do
    @host.expects(:generate_pxe_template).with(:PXELinux).returns("Fake kernel line boot/RedHatEnterpriseLinuxAtomic-7.3-      +x86_64-vmlinuz")
    boot_files = [
      {:"boot/RedHatEnterpriseLinuxAtomic-7.3-x86_64"=>"/tmp/pxeboot/vmlinuz"},
      {:"boot/RedHatEnterpriseLinuxAtomic-7.3-x86_64"=>"/tmp/pxeboot/initrd.img"}
    ]
    @host.operatingsystem.expects(:pxe_files).returns(boot_files)
    ForemanBootdisk::ISOGenerator.expects(:generate).with({
      :isolinux => 'Fake kernel line BOOT/NUXATOMIC_7_3_X86_64_VMLINUZ',
      :files => [[['BOOT/NUXATOMIC_7_3_X86_64_VMLINUZ', '/tmp/pxeboot/vmlinuz']],
        [['BOOT/ATOMIC_7_3_X86_64_INITRD_IMG', '/tmp/pxeboot/initrd.img']]]}, anything)
    ForemanBootdisk::ISOGenerator.generate_full_host(@host)
  end
```

I will file PR once this is merged, filed: http://projects.theforeman.org/issues/18823

**Comment by lzap** [comment]
Thanks here are the missing tests: https://github.com/theforeman/foreman_bootdisk/pull/38
